Build: Shadow x-pack:protocol into x-pack:plugin:core

nik9000 opened the pull request:
This bundles the x-pack:protocol project into the x-pack:plugin:core
project because we'd like folks to consider it an implementation detail
of our build rather than a separate artifact to be managed and depended
on. It is now bundled into both x-pack:plugin:core and
client:rest-high-level. To make this work I had to fix a few things.

Firstly, I had to make PluginBuildPlugin work with the shadow plugin.
In that case we have to bundle only the `shadow` dependencies and the
shadow jar.

Secondly, every reference to x-pack:plugin:core has to use the `shadow`
configuration. Without that the reference is missing all of the
un-shadowed dependencies. I tried to make it so that applying the shadow
plugin automatically redefines the `default` configuration to mirror the
`shadow` configuration which would allow us to use bare project references
to the x-pack:plugin:core project but I couldn't make it work. It'd *look*
like it works but then fail for transitive dependencies anyway. I think
it is still a good thing to do but I don't have the willpower to do it
now.

Finally, I had to fix an issue where Eclipse and IntelliJ didn't properly
reference shadowed transitive dependencies. Neither IDE supports shadowing
natively so they have to reference the shadowed projects. We fix this by
detecting `shadow` dependencies when in "Intellij mode" or "Eclipse mode"
and adding `runtime` dependencies to the same target. This convinces
IntelliJ and Eclipse to play nice.

nik9000 committed:
WIP

nik9000 committed:
Wip

nik9000 committed:
Works

nik9000 committed:
Wip

nik9000 committed:
Closer

nik9000 committed:
woops

nik9000 committed:
More

nik9000 committed:
Moar shadow

nik9000 committed:
Revert bad

nik9000 committed:
Explain

nik9000 committed:
Fix eclipse

nik9000 committed:
Fix intellij

nik9000 committed:
Fix featureAware

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy:
```diff
@@ -153,8 +141,24 @@ public class PluginBuildPlugin extends BuildPlugin {
+            /*
+             * We bundle the plugin's jar file and its dependencies. We have
+             * to wait until the project is evaluated before we know if the
+             * plug *has* the shadow plugin. If it does we need the shadow
```
plug -> project?

nik9000 replied:
++

rjernst reviewed (approved):
This looks fine. A couple general comments:
* Can you add docs somewhere in contributing or something like that on the ability to shade and using the shadow vs compile?
* The name "shadow" for the configuration of what deps are *not* shaded. Is there a way to change this?

nik9000 commented on build.gradle:
```diff
@@ -516,6 +516,32 @@ allprojects {
 
+allprojects {
+  /*
+   * IntelliJ and, ironically, Eclipse don't know about the shadow plugin so
```
Will do. It is ironic because and eclipse is caused by a shadow.

nik9000 commented:
> The name "shadow" for the configuration of what deps are not shaded. Is there a way to change this?

We'd have to make a new configuration. I was weary to do that at first but I think I understand how all things hangs together well enough to do that. What do you think of the name `unshaded`? What do you think of doing it in a follow up PR?

> Can you add docs somewhere in contributing or something like that on the ability to shade and using the shadow vs compile?

I'll add that before merging.

nik9000 commented:
And thanks for reviewing @rjernst!

nik9000 committed:
Merge branch 'master' into x_pack_transport_protocol

nik9000 committed:
Explanation of configurations

Includes a start of explaining gradle and gradle projects which feels
sort of required before talking about configurations. Not at all
finished but a start.

alpar-t reviewed (commented):
What is the advantage of using the shadow plugin as opposed to configuring the x-pack:core, high level rest and possibly other builds to simply include the class and resource files from the protocol project as in ` jar { from project(':x-pack:protocol').classes ; from project(':x-pack:protocol').processResources }` ( or something like that )  with a compile only dependency on `project(':x-pack:protocol')` to keep the IDEs happy about it could do this without leaking shadow plugin semantics across the build. It feels to me that we are throwing out too much of what the plugin does to warrant using it.  
We could even generalize that by creating a configuration of our own `bundled` or `baked` that can have only project level dependencies that behave like mentioned above: added to the jar and as compileOnly deps. Would probably keep things simpler in the gradle side.

nik9000 commented:
> What is the advantage of using the shadow plugin as opposed to configuring the x-pack:core, high level rest and possibly other builds to simply include the class and resource files from the protocol project as in jar { from project(':x-pack:protocol').classes ; from project(':x-pack:protocol').processResources } ( or something like that ) with a compile only dependency on project(':x-pack:protocol') to keep the IDEs happy about it could do this without leaking shadow plugin semantics across the build. It feels to me that we are throwing out too much of what the plugin does to warrant using it.
> We could even generalize that by creating a configuration of our own bundled or baked that can have only project level dependencies that behave like mentioned above: added to the jar and as compileOnly deps. Would probably keep things simpler in the gradle side.

I don't think it is worth re-implementing a bunch of the semantics of the shadow plugin though. We actually activate the META-INF merging across the board which is quite useful. And the jdbc driver uses relocation. So I think the shadow plugin is the right choice for us.

I don't think we can get away without leaking shadow semantics out of the build because we'd end up including the bundled classes twice in dependencies which will cause jar hell to fail. I'd *love* to alter the `default` configuration of a project when it has the shadow plugin so that we can just depend on the project and get the appropriate target. I just couldn't get it to work and got tired of fighting it.

I'll certainly look at your other points!

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy:
```diff
@@ -873,11 +876,20 @@ class BuildPlugin implements Plugin<Project> {
             it.group.startsWith('org.elasticsearch') == false
         } - project.configurations.compileOnly
+        project.plugins.withType(ShadowPlugin).whenPluginAdded {
+            project.dependencyLicenses.dependencies += project.configurations.shadow.fileCollection {
```
It mimics what we do above for the `runtime` configuration. I'm not arguing that it is right, just that it is in good company.

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy:
```diff
@@ -153,8 +141,24 @@ public class PluginBuildPlugin extends BuildPlugin {
+             * we're fine with the jar an all of the non-provided runtime
+             * dependencies.
+             */
+            project.afterEvaluate {
```
That works! Awesome!

nik9000 committed:
Wait until the last second!

nik9000 committed:
Link

nik9000 committed:
Fix last jarHell

nik9000 commented on CONTRIBUTING.md:
```diff
@@ -305,6 +305,39 @@ the `qa` subdirectory functions just like the top level `qa` subdirectory. The
+
+<dl>
+<dt>`compile`</dt><dd>Code that is on the classpath at both compile and
+runtime. If the `shadow` plugin is applied to the project then this code is
```
Linked!

atorok commented:
LGTM  Would be great if we could change the default configuration to shadow  but we can do it in another PR as discussed.

nik9000 commented:
Thanks for reviewing @atorok and @rjernst! I have two follow up things to do after I finish backporting this:

1. Figure out if I can make the `default` context work extend from the `shadow` context so you don't need to say `context: 'shadow'` all over the build.
1. Figure out if I can rename the `shadow` context into something more obvious like `unbundled`.

rjernst commented:
I like the idea of having a `bundled` configuration, but all of these ideas are things that we can iterate on. If this works, let's go with it for now.

nik9000 commented:
I opened #32409 to rework the configuration of the shadow plugin. I think it makes it much more obvious what is going on.
